frontend/src/views/producer/store.js

```diff
@@ -1,63 +1,175 @@
+      fetchError: types.optional(
+        types.union(types.string, types.boolean),
+        false,
+      ),
+    }),
+    {},
+  ),
+});
 
 export const RootProducerModel = types
   .model({
     events: types.optional(types.map(EventModel), {}),
     selectedEvent: types.maybe(types.reference(EventModel)),
+    user: types.optional(User, User.create().toJSON()),
+    ui: types.optional(UI, {}),
   })
   .actions(self => ({
-    addEvent: event => self.events.set(event.id, EventModel.create(event)),
+    // hooks
+    afterCreate() {
+      if (window.localStorage.getItem('store'))
+        applySnapshot(self, JSON.parse(window.localStorage.getItem('store')));
+      if (self.events[self.selectedEvent]) self.selectedEvent = null;
+      self.fetchEvents();
+    },
+    // event actions
     selectEvent: id => (self.selectedEvent = id),
-    patchEvent: event => {
-      self.events.set(event.id, event);
+    fetchEvents: flow(function*() {
+      try {
+        self.ui.eventList.fetching = true;
+        const response = yield window.fetch(BASE_URL + '/events');
+        if (!response.ok)
+          throw new Error('Failed to fetch. Server threw HTTP error code');
+        const json = yield response.json();
+        const processedJson = pipeable(json).pipe(
+          parseTo('camelCase'),
+          transformToMap,
+        );
+        self.ui.eventList.fetching = false;
+        applySnapshot(self.events, processedJson);
+        self.selectedEvent = null;
+      } catch (error) {
+        console.error(error);
+        self.ui.eventList.fetching = false;
+        self.ui.eventList.fetchError = 'Could not fetch events ';
+      }
+    }),
+    patchEvent: flow(function*(event) {
+      if (!resolveIdentifier(EventModel, self.events, event.id)) {
+        // new event. POST
+        try {
+          self.ui.eventList.fetching = true;
+          const response = yield window.fetch(BASE_URL + '/events', {
+            headers: {
+              'Content-Type': 'application/json',
+              Authorization: `Bearer ${self.user.token}`,
+            },
+            method: 'POST',
+            body: pipeable(event).pipe(
+              parseTo('snake'),
+              removeId,
+              JSON.stringify,
+            ),
+          });
+          if (!response.ok)
+            throw new Error('Failed to fetch. Server threw HTTP error code');
+          const json = yield response.json();
+          const processedJson = parseTo('camelCase')(json);
+
+          self.ui.eventList.fetching = false;
+          self.events.put(processedJson);
+          self.selectEvent(processedJson.id);
+        } catch (error) {
+          console.error(error);
+          self.ui.eventList.fetching = false;
+          self.ui.eventList.fetchError = 'Could not create event ' + event.name;
+        }
+      } else {
+        // existing event. PUT
+        try {
+          self.ui.eventList.fetching = true;
+          const response = yield window.fetch(
```

Comment:
How about making a helper function for making these fetches, so that you wouldn't need to repeat the headers every time? Also in that function you could handle the eventList.fetching flag, since you seem to be repeating that one also quite many times..

Reply:
i think i will pull these to a "api" sections that deals with the endpoints

Reply:
👍